zod/src/__tests__/zod.ts

```diff
@@ -89,4 +91,86 @@ describe('zodResolver', () => {
 
     await expect(promise).rejects.toThrow('custom error');
   });
+
+  /**
+   * Type inference tests
+   */
+  it('should correctly infer the output type from a zod schema', () => {
+    const resolver = zodResolver(z.object({ id: z.number() }));
+
+    expectTypeOf(resolver).toEqualTypeOf<
+      Resolver<FieldValues, any, { id: number }>
+    >();
+  });
+
+  it('should correctly infer the output type from a zod schema when a different input type is specified', () => {
+    const schema = z.object({ id: z.string() });
+    const resolver = zodResolver<{ id: string }, any, z.output<typeof schema>>(
+      schema,
+    );
+
+    expectTypeOf(resolver).toEqualTypeOf<
+      Resolver<{ id: string }, any, { id: string }>
+    >();
+  });
+
+  it('should correctly infer the output type from a zod schema when a different input type is specified (only input type is specified)', () => {
+    const resolver = zodResolver<{ id: string }>(z.object({ id: z.string() }));
+
+    expectTypeOf(resolver).toEqualTypeOf<
+      Resolver<{ id: string }, any, { id: string }>
+    >();
+  });
+
+  it('should correctly infer the output type from a zod schema when different input and output types are specified', () => {
+    const resolver = zodResolver<
+      { id: string },
+      { context: any },
+      { id: boolean }
+    >(z.object({ id: z.number() }));
+
+    expectTypeOf(resolver).toEqualTypeOf<
+      Resolver<{ id: string }, { context: any }, { id: boolean }>
+    >();
+  });
+
+  it('should correctly infer the output type from a Zod schema for the handleSubmit function in useForm', () => {
+    const { handleSubmit } = useForm({
+      resolver: zodResolver(z.object({ id: z.number() })),
+    });
+
+    expectTypeOf(handleSubmit).parameter(0).toEqualTypeOf<
+      SubmitHandler<{
+        id: number;
+      }>
+    >();
+  });
+
+  it('should correctly infer the output type from a Zod schema when a different input type is specified for the handleSubmit function in useForm', () => {
+    const { handleSubmit } = useForm({
+      resolver: zodResolver<{ id: number }>(z.object({ id: z.string() })),
+    });
+
+    expectTypeOf(handleSubmit).parameter(0).toEqualTypeOf<
+      SubmitHandler<{
+        id: string;
+      }>
+    >();
+  });
+
+  it('should correctly infer the output type from a Zod schema when different input and output types are specified for the handleSubmit function in useForm', () => {
+    const resolver = zodResolver<{ id: string }, any, { id: boolean }>(
+      z.object({ id: z.number() }),
+    );
```

Comment:
It seems there is a simple mistake here. should be:

```suggestion
    const resolver = zodResolver<{ id: string }, any, { id: boolean }>(
      z.object({ id: z.string() }),
    );
```

Additionally, since the `Input` type of the generic parameters and the `Schema` type don't match, a type error should actually occur here. This would indicate that the code is not type-safe.  

To fix this, you should pass the `Output` and `Input` types to the `ZodSchema` like so:

```ts
export function zodResolver<
  Input extends FieldValues,
  Context,
  Output,
  Schema extends z.ZodSchema<Output, any, Input> = z.ZodSchema<Output, any, Input>
>(
  //...
);
```

Reply:
thanks @kotarella1110 and 👋

Reply:
Fixed ✅